client/src/main/java/org/apache/celeborn/client/read/LocalPartitionReader.java

```diff
@@ -0,0 +1,210 @@
+    this.location = location;
+    this.endMapIndex = endMapIndex;
+    PbStreamHandler streamHandle;
+    long fetchTimeoutMs = conf.clientFetchTimeoutMs();
+    try {
+      TransportClient client =
+          clientFactory.createClient(location.getHost(), location.getFetchPort(), 0);
+      TransportMessage openStreamMsg =
+          new TransportMessage(
+              MessageType.OPEN_STREAM,
+              PbOpenStream.newBuilder()
+                  .setShuffleKey(shuffleKey)
+                  .setFileName(location.getFileName())
+                  .setStartIndex(startMapIndex)
+                  .setEndIndex(endMapIndex)
+                  .setReadLocalShuffle(true)
+                  .build()
+                  .toByteArray());
+      ByteBuffer response = client.sendRpcSync(openStreamMsg.toByteBuffer(), fetchTimeoutMs);
+      streamHandle = TransportMessage.fromByteBuffer(response).getParsedPayload();
+    } catch (IOException | InterruptedException e) {
+      throw new IOException(
+          "Read shuffle file from local file failed, partition location: "
+              + location
+              + " filePath: "
+              + location.getStorageInfo().getFilePath(),
+          e);
+    }
+
+    chunkOffsets = new ArrayList<>(streamHandle.getChunkOffsetsList());
+    numChunks = streamHandle.getNumChunks();
+    shuffleChannel = FileChannelUtils.openReadableFileChannel(streamHandle.getFullPath());
+
+    logger.debug(
+        "Local partition reader {} offsets:{}",
+        location.getStorageInfo().getFilePath(),
+        StringUtils.join(chunkOffsets, ","));
+
+    if (hasPendingFetchTask.compareAndSet(false, true)) {
+      Future firstFetchTask = readLocalShufflePool.submit(() -> fetchDataTask());
+      try {
+        firstFetchTask.get();
+      } catch (Exception e) {
+        throw new CelebornIOException("Fetch data for local shuffle reader failed", e);
+      }
+    }
+    ShuffleClient.incrementLocalReadCounter();
+  }
+
+  private void fetchDataTask() {
+    try {
+      if (!closed && currentChunkIndex < numChunks && results.size() < fetchMaxReqsInFlight) {
+        long offset = chunkOffsets.get(currentChunkIndex);
+        long length = chunkOffsets.get(currentChunkIndex + 1) - offset;
+        logger.debug("Read {} offset {} length {}", currentChunkIndex, offset, length);
+        // A chunk must be smaller than INT.MAX_VALUE
+        ByteBuffer buffer = ByteBuffer.allocate((int) length);
+        if (endMapIndex != Integer.MAX_VALUE) {
+          // skew partition
+          shuffleChannel.position(offset);
+        }
+        while (buffer.hasRemaining()) {
+          if (-1 == shuffleChannel.read(buffer)) {
+            exception.set(
+                new CelebornIOException(
+                    "Read local file " + location.getStorageInfo().getFilePath() + " failed"));
+          }
+        }
+        buffer.flip();
+        results.put(Unpooled.wrappedBuffer(buffer));
```

Comment:
We should check `closed` before adding into `results` in a synchronized block, as `WorkerPartitionReader` does. I think this file should follow `WorkerPartitionReader`'s structure since it's fully verified.

Reply:
Closed in `WorkerPartitionReader` is not volatile.

Reply:
> Closed in `WorkerPartitionReader` is not volatile.

It's not about volatile, it's about resource leak. Consider this situation:
1. At time 1 thread 1 runs `fetchDataTask`, at this time closed is false
2. At time 2 thread 2 calls `close`, sets `closed` true, and releases `results`.
3. At time 3 thread 1 puts buffer into `results`. This may cause resource leak.